File: LivrusVirtus/BoaConstructor/wxPython/RotinasGlobais.py
from wxPython.wx import *
from time import localtime, strftime, strptime
from resource import Resource
from ListaStr import LISTAMSTR
import fpformat

def create():
  return RotinasGlobais()

class RotinasGlobais:
  gConexao = None
  sBarraLogin = ''
  sUsuarioLogin = ''
  sUsuarioSenha = ''
  sCodigoSelecionado = ''
  sPreco = ''
  sQtdEstoque = ''
  sepDTHR = '#' #SOMENTE ACCESS
  #sepDTHR = "'" #OUTROS BANCOS

  def __init__(self):
    self.data = []

  def ValidaFloat(self, sNum):
    nNum = 0.0
    bret = False

    if len(sNum) > 0:
      try:
        nNum = float(sNum)
        if nNum >= 0.01:
          bret = True
      except:
        wxMessageBox(Resource.STR_NUMINV, u'Erro', wxOK | wxICON_ERROR)
        bret = False
    else:
      bret = False
    
    return bret
  
  def VirgulaParaPonto(self, sNum, bRevertido):
    if bRevertido:
      sret = sNum.replace('.', ',')
    else:
      sret = sNum.replace(',', '.')
    
    return sret

  def substr(self, stexto, nini, ntam):
    return stexto[nini-1:nini+ntam-1]

  def modulo11(self, svalor, ndig):
    ncont = 1
    slocal = ''
    while (ncont <= 11):
      slocal=self.substr(svalor,ncont,1) + slocal
      ncont=ncont+1
    nsoma=0;
    if (ndig == 1):
      ncont=3;
    else:
      ncont=2;
    npos = 2
    while (ncont <= 11):
      sres = self.substr(slocal,ncont,1)
      nvalor = int(sres)
      nsoma=nsoma + (nvalor * npos)
      ncont=ncont+1
      npos=npos+1
  
    nsoma=(11 - (nsoma % 11))
    if (nsoma > 9):
      return 0
    else:
      return nsoma

  def validaCPF(self, cpf):
    if len(cpf)==14:
      try:
        scpf=self.substr(cpf,1,3)+self.substr(cpf,5,3)+self.substr(cpf,9,3)+\
          self.substr(cpf,13,2)
        sres=self.substr(cpf,13,1)
        nvalor1 = int(sres)
        sres=self.substr(cpf,14,1)
        nvalor2 = int(sres)
        if (self.modulo11(scpf,1)==nvalor1) and\
           (self.modulo11(scpf,2)==nvalor2):
          return True
        else:
          return False
      except:
        return False
    else:
      return False

  def validaISBN(self, isbn):
    if len(isbn)==13:
      sdigito = self.substr(isbn, len(isbn), 1)
      sisbn=''
      for ncont in range(1, len(isbn)):
        if self.substr(isbn,ncont,1)!='-':
          sisbn=sisbn+self.substr(isbn,ncont,1)
  
      if sdigito == 'X':
        nsoma = 10
      else:
        nvalor = int(sdigito)
        nsoma = nvalor
  
      for ncont in range(1, 10):
        snum = self.substr(sisbn, ncont, 1)
        nvalor = int(snum)
        nsoma=nsoma + (nvalor * (11 - ncont))
  
      if ((nsoma % 11) == 0):
        return True
      else:
        return False
    else:
      return False

  def validaDataHora(self, sDataHora, bSoData):
    bvalida=False
    bdatavalida=False
    bhoravalida=False
    nudiames=0
    if bSoData==True: ntam=10
    else: ntam=19

    if len(sDataHora)==ntam:
      sdia = self.substr(sDataHora,1,2)
      ndia = int(sdia)
      smes = self.substr(sDataHora,4,2)
      nmes = int(smes)
      sano = self.substr(sDataHora,7,4)
      nano = int(sano)
    
      if (nano >= 1) and (nano <= 9999):
        if (nmes >= 1) and (nmes <= 31):
            if nmes in [1, 3, 5, 7, 8, 10, 12]: 
              nudiames = 31
            elif nmes == 2: 
              # ano bissexto 
              if (nano % 4) == 0: 
                nudiames = 29
              else: nudiames = 28
            elif nmes in [4, 6, 9, 11]: 
              nudiames = 30
            if (ndia >=1) and (ndia <=nudiames):
               bdatavalida = True
    
      if bSoData:
         if bdatavalida: bvalida = True
         else: bvalida = False
      else:
        if bdatavalida:
            shora = self.substr(sDataHora,12,2)
            nhora = int(shora)
            smin = self.substr(sDataHora,15,2)
            nmin = int(smin)
            sseg = self.substr(sDataHora,18,2)
            nseg = int(sseg)
        
            if (nhora >= 0) and (nhora <= 23):
              if (nmin >= 0) and (nmin <= 59):
                if (nseg >= 0) and (nseg <= 59):
                   bhoravalida = True
        
            if bhoravalida: bvalida = True
            else: bvalida = False
        else:
          bvalida = False
    else:
      bvalida = False
  
    return bvalida

  def retDataHoraStr(self, bData, bHora):
    sDataHora = ''
    
    if bData and bHora:
      sDataHora = strftime('%d/%m/%Y %H:%M:%S', localtime())
    elif bData:
      sDataHora = strftime('%d/%m/%Y', localtime())
    elif bHora:
      sDataHora = strftime('%H:%M:%S', localtime())
    
    return sDataHora;

  def ConverteDataHoraStr(self, sDTHR, bDMY_ou_YMD, sFrmtData, sFrmtHora):
    sDataHora = ''
    sFormato = ''
    
    if len(sFrmtData) > 0:
      if bDMY_ou_YMD:
        ndia = self.substr(sDTHR, 1, 2)
        nmes = self.substr(sDTHR, 4, 2)
        nano = self.substr(sDTHR, 7, 4)
      else:
        nano = self.substr(sDTHR, 1, 4)
        nmes = self.substr(sDTHR, 6, 2)
        ndia = self.substr(sDTHR, 9, 2)
      sDataHora = ndia + '/' + nmes + '/' + nano
      sFormato = '%d/%m/%Y'

    if len(sFrmtHora) > 0:
      nhora = self.substr(sDTHR, 12, 2)
      nmin = self.substr(sDTHR, 15, 2)
      nseg = self.substr(sDTHR, 18, 2)
      sDataHora = sDataHora + nhora + ':' + nmin + ':' + nseg
      sFormato = sFormato + '%H:%M:%S'

    sDTHR = strptime(sDataHora, sFormato)
    sDataHora = strftime( sFrmtData + ' ' + sFrmtHora , sDTHR)
    return sDataHora

  def ConectaBanco(self, sBancoDeDados, sLogin, sSenha):
    import odbc
    #import MySQLdb
    try:
      #self.gConexao = MySQLdb.connect( host = 'localhost', 
      #  db = sBancoDeDados, user = sLogin, passwd = sSenha )
      self.gConexao = odbc.odbc( sBancoDeDados )
    except:
      wxMessageBox(Resource.STR_NCONBANCO, 'Erro', wxOK | wxICON_ERROR)
      return False
    else:
      return True
  
  def DisconectaBanco(self):
    pass
  
  def AtualizaDados(self, hBancoLocal, sSql):
    try:
      cursor = hBancoLocal.cursor()
      # Necessario codificacao de acordo com o banco
      cursor.execute( sSql.encode('latin-1') )   
    except:
      wxMessageBox(Resource.STR_NCEXECSQL, 'Erro', wxOK | wxICON_ERROR)
      return False
    else:
      return True
  
  def ConsultaDados(self, hBancoLocal, sSql):
    try:
      cursor = hBancoLocal.cursor()
      cursor.execute( sSql )   
    except:
      wxMessageBox(Resource.STR_NCEXECSQL, 'Erro', wxOK | wxICON_ERROR)
      return False
    else:
      LISTAMSTR.col = cursor.fetchall()
      LISTAMSTR.tam = len(LISTAMSTR.col)
      if LISTAMSTR.tam > 0:
        return True
      else:
        return False

  def arrumaColunas(self, lstLocal):
    lstLinhas = []
    nQtdCols = len(lstLocal[0])
    for nLinha in range(0,len(lstLocal)):
      lstCols = []
      for nCol in range(0,nQtdCols):
        if type(lstLocal[nLinha][nCol]) is str:
          lstCols.append(lstLocal[nLinha][nCol]) 
        elif type(lstLocal[nLinha][nCol]) is int:
          lstCols.append(str(lstLocal[nLinha][nCol])) 
        elif type(lstLocal[nLinha][nCol]) is float:
          lstCols.append(fpformat.fix(lstLocal[nLinha][nCol],2)) 
        else:
          sDTHR = strptime( str(lstLocal[nLinha][nCol]), "%a %b %d %H:%M:%S %Y")
          lstCols.append(strftime( '%d/%m/%Y %H:%M:%S' , sDTHR)) 
      
      lstLinhas.append(lstCols) 
  
    return lstLinhas

  def ConsultaDadosLista(self, hBancoLocal, lstLocal, sSql):
    try:
      cursor = hBancoLocal.cursor()
      cursor.execute( sSql )   
    except:
      wxMessageBox(Resource.STR_NCEXECSQL, 'Erro', wxOK | wxICON_ERROR)
      return False
    else:
      lstLocal.col = cursor.fetchall()
      lstLocal.tam = len(lstLocal.col)
      if lstLocal.tam > 0:
        lstLocal.col = self.arrumaColunas(lstLocal.col)
        return True
      else:
        return False

  def ValidaCampo(self, sCampo, sMensagemErro):
    bret = False
    if len(sCampo)==0:
      wxMessageBox(sMensagemErro, "Erro", wxOK | wxICON_ERROR)
      bret = True
    else:
  	  bret = False
  
    return bret

  def ConsultaCampoDesc(self, Campos, sTextoSql, sMensagemErro):
    sret = ""
    if self.ConsultaDadosLista(self.gConexao, Campos, sTextoSql):
      sret = Campos.col[0][1]
    else:
      wxMessageBox(sMensagemErro, "Erro", wxOK | wxICON_ERROR)
      sret = ""
    
    return sret

  def AdicionaItemLista(self, lstLista, slCodLista, sCampoCodigo, sCampoDesc):
    sItems = ''; sCodigo = ''; sDesc = ''
    
    if len(sCampoDesc.GetValue()) > 0:
      sCodigo = sCampoCodigo.GetValue()
      sDesc = sCampoDesc.GetValue()
      
      sItems = sCodigo + ' - ' + sDesc
      
      if lstLista.FindString(sItems) == -1:
        slCodLista.col.append(sCodigo)
        lstLista.Append(sItems)
    
      sCampoCodigo.Clear()
      sCampoDesc.Clear() 

  def RemoveItemLista(self, lstLista, slCodLista):
    nSelIndice = 0 
    if lstLista.GetCount() > 0:
      nSelIndice = lstLista.GetSelection()
      if lstLista.Selected(nSelIndice):
        slCodLista.col.pop(nSelIndice);
        lstLista.Delete(nSelIndice)

  def AdicionaValoresLista(self, sTextoSql, lstLista, slCodLista):
    Campos = LISTAMSTR()
    sItems = ''; sCodigo = ''; sDesc = ''
    nLinha = 0
  
    lstLista.Clear()
    slCodLista.col = []
    if self.ConsultaDadosLista(self.gConexao, Campos, sTextoSql): 
      for nLinha in range(0, len(Campos.col)):
        sCodigo = Campos.col[nLinha][0]
        sDesc = Campos.col[nLinha][1]
  
        slCodLista.col.append(sCodigo)
        sItems = sCodigo + ' - ' + sDesc
        lstLista.Append(sItems)

  def AdicionaColunasGrid(self, grid, nPosIni, nQtdCols, nLargura, sRotulos):
    nPos = 0; nCont = 0
    
    if nPosIni > 0:
      grid.AppendCols(nQtdCols)
    nCont = 0
    for nPos in range(nPosIni, nPosIni+nQtdCols):
      grid.SetColSize(nPos, nLargura[nCont])
      grid.SetColLabelValue(nPos, sRotulos[nCont])
      nCont = nCont + 1

  def AdicionaLinhasGrid(self, grid, lstLinhas, nQtdCol):
    nCol = 0; nLinha = 0
  
    grid.DeleteRows(0, grid.GetNumberRows(), True)
    grid.AppendRows(lstLinhas.tam, True)
    for nCol in range(0, nQtdCol):
      for nLinha in range(0, lstLinhas.tam):
        grid.SetCellValue(nLinha, nCol, str(lstLinhas.col[nLinha][nCol]))

File: LivrusVirtus/wxWindows/wxRuby/RotinasGlobais.rb
require 'wxruby'
require 'DBI'
# require 'DBD::ODBC'

require 'resource'
require 'ListaStr'

$Rotinas = nil

class RotinasGlobais
  attr :gConexao, true
  attr :sBarraLogin, true
  attr :sUsuarioLogin, true
  attr :sUsuarioSenha, true
  attr :sCodigoSelecionado, true
  attr :sPreco, true
  attr :sQtdEstoque, true
  SepDTHR = "#" #SOMENTE ACCESS
  #SepDTHR = "'" #OUTROS BANCOS

  def MessageErro(sTexto)
    mens = Wx::MessageDialog.new(nil, sTexto, 'Erro', Wx::OK | Wx::ICON_ERROR)
    mens.show_modal()
  end

  def ValidaFloat(sNum)
    if sNum != ""
      if sNum.to_f > 0
        return true
      else
        MessageErro(Resource::STR_NUMINV)
        return false
      end
    else
     return false
    end
  end 

  def VirgulaParaPonto(sNum, bRevertido)
    if (bRevertido)
      return sNum.sub(/[.]/, ',')
    else
      return sNum.sub(/[,]/, '.')
    end
  end

  def modulo11(svalor, ndig)
    ncont, nsoma, npos, nvalor = 0,0,0,0
    sres, slocal = "", ""
  
    ncont=1
    while ncont <= 11 do
      slocal=svalor.slice(ncont-1,1) + slocal
      ncont=ncont+1
    end
    nsoma=0
    if ndig == 1 then
      ncont=3
    else
      ncont=2
    end
    npos=2
    while ncont <= 11 do
      sres=slocal.slice(ncont-1,1)
      nvalor = sres.to_i
      nsoma=nsoma + (nvalor * npos)
      ncont=ncont+1
      npos=npos+1
    end
    nsoma=(11 - (nsoma % 11))
    
    if nsoma > 9
      return 0
    else
      return nsoma
    end
  end

  def validaCPF(cpf)
    scpf, sres = "", ""
    nvalor1, nvalor2 = 0,0
    
    if cpf.length==14 then
      scpf=cpf.slice(0,3)+cpf.slice(4,3)+cpf.slice(8,3)+cpf.slice(12,2)
      sres=cpf.slice(12,1)
      nvalor1 = sres.to_i
      sres=cpf.slice(13,1)
      nvalor2 = sres.to_i
      if (modulo11(scpf,1)==nvalor1) and (modulo11(scpf,2)==nvalor2)
        return true
      else
        return false
      end
    else
      return false
    end
  end

  def validaISBN(isbn)
    snum, sdigito, sisbn = "", "", ""
    ncont, nsoma, nvalor = 0,0,0
  
    if isbn.length==13 then
      sdigito=isbn.slice(isbn.length-1, 1)
      sisbn = ""
      for ncont in (1..isbn.length-1) do
        if isbn.slice(ncont-1,1)!="-" then
          sisbn=sisbn+isbn.slice(ncont-1,1)
        end
      end
      if sdigito == "X" then
        nsoma = 10
      else
        nvalor = sdigito.to_i
        nsoma = nvalor
      end
      for ncont in (1..9) do
        snum=sisbn.slice(ncont-1, 1)
        nvalor = snum.to_i
        nsoma=nsoma + (nvalor * (11 - ncont))
      end
      if (nsoma % 11) == 0
        return true
      else
        return false
      end
    else
      return false
    end
  end

  def validaDataHora(sDataHora, bSoData)
    sdia, smes, sano, shora, smin, sseg = "","","","","",""
    ndia, nmes, nano, nhora, nmin, nseg, ntam = 0,0,0,0,0,0,0 
    
    bvalida=false
    bdatavalida=false
    bhoravalida=false
    nudiames=0
    
    if bSoData==true then ntam=10
    else ntam=19
    end
    if sDataHora.length==ntam then
      sdia=sDataHora.slice(0,2)
      ndia=sdia.to_i
      smes=sDataHora.slice(3,2)
      nmes=smes.to_i
      sano=sDataHora.slice(6,4)
      nano=sano.to_i
    
      if (nano >= 1) and (nano <= 9999) then
        if (nmes >= 1) and (nmes <= 31) then
            case nmes
              when 1, 3, 5, 7, 8, 10, 12
                nudiames=31
              when 2
                 # ano bissexto
                 if (nano % 4)==0 then 
                   nudiames=29
                 else 
                   nudiames=28
                 end
               when 4, 6, 9, 11
                 nudiames=30
            end
            if (ndia >=1) and (ndia <=nudiames) then
               bdatavalida=true
            end 
        end
      end
    
      if bSoData then
         if bdatavalida then 
           bvalida=true
         else 
           bvalida=false
         end 
      else
        if bdatavalida then
            shora=sDataHora.slice(11,2)
            nhora = shora.to_i
            smin=sDataHora.slice(14,2)
            nmin = smin.to_i
            sseg=sDataHora.slice(17,2)
            nseg = sseg.to_i
    
            if (nhora >= 0) and (nhora <= 23) then
              if (nmin >= 0) and (nmin <= 59) then
                if (nseg >= 0) and (nseg <= 59) then
                   bhoravalida=true
                end
              end
            end
    
            if bhoravalida then 
               bvalida=true
            else 
               bvalida=false
            end
        else
          bvalida=false
        end
      end
    else
      bvalida=false
    end
    return bvalida
  end

  def retDataHoraStr(bData, bHora)
    sDataHora = ''
    sTempo = Time.now 
    if (bData) and (bHora) then
      sDataHora = sTempo.strftime("%d/%m/%Y %H:%M:%S")
    elsif bData then
      sDataHora = sTempo.strftime("%d/%m/%Y")
    elsif bHora then
      sDataHora = sTempo.strftime("%H:%M:%S")
    end
    return sDataHora
  end

  def ConverteDataHoraStr(sDTHR, bDMY_ou_YMD, sFrmtData, sFrmtHora)
    sDataHora = ''
    
    if sFrmtData.length > 0 then
      if bDMY_ou_YMD then
        ndia = sDTHR.slice(0, 2)
        nmes = sDTHR.slice(3, 2)
        nano = sDTHR.slice(6, 4)
      else
        nano = sDTHR.slice(0, 4)
        nmes = sDTHR.slice(5, 2)
        ndia = sDTHR.slice(8, 2)
      end
    end
    if sFrmtHora.length > 0 then
      nhora = sDTHR.slice(11, 2)
      nmin = sDTHR.slice(14, 2)
      nseg = sDTHR.slice(17, 2)
    end
    sTempo = Time.local(nano, nmes, ndia, nhora, nmin, nseg, 0)
    if sFrmtHora.length > 0 then
      sDataHora = sTempo.strftime( sFrmtData + ' ' + sFrmtHora )
    else
      sDataHora = sTempo.strftime( sFrmtData + ' 00:00:00' )
    end
    return sDataHora
  end

  def ConectaBanco(sBancoDeDados, sLogin, sSenha)
    begin
      self.gConexao = DBI.connect( sBancoDeDados, sLogin, sSenha ) 
      return true
    rescue
      MessageErro(Resource::STR_NCONBANCO)
      return false
    end
  end
  
  def DisconectaBanco()
    self.gConexao.disconnect()
  end

  def AtualizaDados(hBancoLocal, sSql)
    begin
      sth = hBancoLocal.prepare( sSql )
      sth.execute()
      sth.finish()
      return true
    rescue
      MessageErro(Resource::STR_NCEXECSQL)
      return false
    end
  end

  def ConsultaDados(hBancoLocal, sSql)
    sth = hBancoLocal.prepare( sSql )
    rc = sth.execute()
    if rc then
      $ListaMStr.col = sth.fetch
      if $ListaMStr.col then
        $ListaMStr.tam = $ListaMStr.col.length
        sth.finish()
        if $ListaMStr.tam > 0 then
          return true
        else
          return false
        end
      else
        return false
      end
    else
      MessageErro(Resource::STR_NCEXECSQL)
      sth.finish()
      return false
    end
  end

  def arrumaColunas(sth)
    lstLinhas = []
    nCol = 0
    nLinha = 0
    while sDBILinha = sth.fetch do
      lstCols = []		
      for nCol in (0..sDBILinha.length-1) do
        if sDBILinha[nCol].class == String then
          lstCols[nCol] = sDBILinha[nCol]
        elsif sDBILinha[nCol].class == Fixnum then
          lstCols[nCol] = sDBILinha[nCol].to_s
        elsif sDBILinha[nCol].class == Float then
          lstCols[nCol] = "%.2f" % sDBILinha[nCol]
        else
          lstCols[nCol] = sDBILinha[nCol].to_time.strftime('%d/%m/%Y %H:%M:%S')
        end
      end
      lstLinhas[nLinha] = lstCols
      nLinha+=1
    end
    return lstLinhas
  end

  def ConsultaDadosLista(hBancoLocal, lstLocal, sSql)
    sth = hBancoLocal.prepare( sSql )
    rc = sth.execute()
    if rc then
      lstLocal.col = arrumaColunas(sth)
      if lstLocal.col then
        lstLocal.tam = lstLocal.col.length     
        if lstLocal.tam > 0 then
          sth.finish()
          return true
        end
      else
        return false
      end
    else
      MessageErro(Resource::STR_NCEXECSQL)
      sth.finish()
      return false
    end
  end

  def ValidaCampo(sCampo, sMensagemErro)
    if sCampo.strip == "" then
      MessageErro(sMensagemErro)
      return true
    else
      return false
    end
  end

  def ConsultaCampoDesc(lstCampos, sTextoSql, sMensagemErro)
    sret = ""
    if ConsultaDadosLista(self.gConexao, lstCampos, sTextoSql) then
      sret = lstCampos.col[0][1]
    else
      MessageErro(sMensagemErro)
      sret = ""
    end
    return sret
  end

  def PesquisaItemLista(lstLista, sItem)
    bret = -1
    for nLinha in (0..lstLista.get_count()-1) do
      if lstLista.get_string(nLinha) == sItem then
        bret = 0
      end
    end
    return bret
  end

  def AdicionaItemLista(lstLista, slCodLista, sCampoCodigo, sCampoDesc)        
    if sCampoDesc.get_value().length > 0 then
      sCodigo = sCampoCodigo.get_value()
      sDesc = sCampoDesc.get_value()
      
      sItems = sCodigo + ' - ' + sDesc
      
      if PesquisaItemLista(lstLista, sItems) == -1 then
        slCodLista.col.push(sCodigo)
        lstLista.append(sItems)
      end
      sCampoCodigo.clear()
      sCampoDesc.clear()
    end 
  end

  def RemoveItemLista(lstLista, slCodLista)
    nSelIndice = 0 
    if lstLista.get_count() > 0 then
      nSelIndice = lstLista.get_selection()
      if lstLista.is_selected(nSelIndice) then
        slCodLista.col.delete_at(nSelIndice)
        lstLista.delete(nSelIndice)
      end
    end
  end

  def AdicionaValoresLista(sTextoSql, lstLista, slCodLista)
    lstCampos = LISTAMSTR.new
    nLinha = 0
  
    lstLista.clear()
    slCodLista.col = []
    if ConsultaDadosLista(self.gConexao, lstCampos, sTextoSql) then
      for nLinha in (0..lstCampos.tam-1) do
        sCodigo = lstCampos.col[nLinha][0]
        sDesc = lstCampos.col[nLinha][1]
  
        slCodLista.col.push(sCodigo)
        sItems = sCodigo + ' - ' + sDesc
        lstLista.append(sItems)
      end
    end
  end

  def AdicionaColunasGrid(grid, nPosIni, nQtdCols, nLargura, sRotulos)
    nPos = 0 
    
    if nPosIni > 0 then
      grid.append_cols(nQtdCols)
    end
    nCont = 0
    for nPos in (nPosIni..(nPosIni+nQtdCols-1)) do
      grid.set_col_size(nPos, nLargura[nCont])
      grid.set_col_label_value(nPos, sRotulos[nCont])
      nCont = nCont + 1
    end
  end

  def AdicionaLinhasGrid(grid, lstLinhas, nQtdCol)
    nCol = 0
    nLinha = 0

    grid.delete_rows(0, grid.get_number_rows(), true)
    grid.append_rows(lstLinhas.tam, true)
    for nCol in (0..nQtdCol-1) do
      for nLinha in (0..lstLinhas.tam-1) do
        grid.set_cell_value(nLinha, nCol, lstLinhas.col[nLinha][nCol])        
      end
    end
  end

end
